pkg/events/event_ringbuffer.go

```diff
@@ -0,0 +1,202 @@
+/*
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package events
+
+import (
+	"sync"
+
+	"github.com/apache/yunikorn-scheduler-interface/lib/go/si"
+)
+
+type resultState int
+
+const (
+	resultOK resultState = iota
+	bufferEmpty
+	idNotFound
+)
```

Comment:
I think `resultState` makes things complex for clients. Why?

resultOK : Normal return status.
bufferEmpty: This can be represented as an empty result (which is a lot more useful).
idNotFound: Ok, now what? Client has to request an ID that is within the buffer, but it doesn't know which one to use.

I would change the structure of GetEventsFromID to the following:

`func (e *eventRingBuffer) GetEventsFromID(id uint64, count unit64) ([]*si.EventRecord, startID uint64)`

Let `startID` be the ID of the first result returned. If the result is empty, let `startID` be the first ID that should be requested on subsequent calls. So in the empty case, this returns `{ [], 0 }`. For the existing `idNotFound` case, if the ID presented is too low, simply start from the real lowest ID and return the first `count` results (or less if there aren't that many). The fact that `startID` in the result didn't equal the ID requested by the client is proof enough that an overflow occurred. If the ID is too high, return an empty list along with the lowest valid starting ID. Again, this is enough to tell the client that the range it requested was not valid, and also gives the startID that should be requested next. In all cases, the returned `startID` + length of the returned list gives the ID that should be requested for the next page.

We should also add `count` as otherwise, the entire buffer may be returned, so we can't paginate properly.

With these changes, I also think we can remove `GetRecentEntries` as this is largely redundant. Newly started clients will likely want to retrieve all history they are not aware of (in which case calling `GetEventsFromID(0, pageSize)` will return the first set of results) or all history beyond a known state (in which case calling `GetEventsFromID(lastKnownEventID+1, pageSize)` will do the job. For completeness we could add `GetLatestEventID() unit64` which would simply return `id`.

Reply:
Done.

I kept `GetRecentEntries()` for the time being, so that Wilfred can share his opinion too. This affects the design described in the document.